### PR TITLE
Cleanup pool types

### PR DESCRIPTION
**Pull request by gruberb**
# Description

Moving out types from `pool-registry` and `pool-system` into `lib/types/src/...`. We then can use the types throughout the codebase through `cfg_types::...`.

`tree libs/types`:

```
├── Cargo.toml
└── src
    ├── adjustments.rs
    ├── epoch.rs
    ├── fee_keys.rs
    ├── fixed_point.rs
    ├── fulfillments.rs
    ├── ids.rs
    ├── investments.rs
    ├── lib.rs
    ├── orders.rs
    ├── permissions.rs
    ├── pools.rs
    ├── reserves.rs
    ├── time.rs
    ├── tokens.rs
    ├── tranches.rs
    └── xcm.rs
```

For example, instead of:
`use cfg_types::CurrencyId`

we now do:
`use cfg_types::tokens::CurrencyId`.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

All current tests run through.

# Checklist:

- [ ] I have added Rust doc comments to structs, enums, traits and functions
- [ ] I have made corresponding changes to the documentation
- [x] I have performed a self-review of my code
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I rebased on the latest `main` branch

### CODE REVIEW

**Comment by mustermeiszer** [comment]
> However, I can't seem to build it with that feature flag. I removed it for now to validate the tests, and have to figure out now how we can, in a test scenario, pass the `test` flag to `cfg-types` and build it with the two additionals `impl` blocks, which are just used for testing.

You must create a feature flag like `std` and pass it to the dependency. Like `with_test_impls` and then we can hide them that way.

**Comment by gruberb** [comment]
> You must create a feature flag like `std` and pass it to the dependency. Like `with_test_impls` and then we can hide them that way.

Wouldn't this alter our `test` command
`cargo test --workspace --release --features test-benchmarks,try-runtime,runtime-benchmarks`? 

I would need one of `test` or the feature flags being passed down and add this flag on top of the `impl` block.

File: libs/types/src/consts.rs [hunk 1]
```diff
@@ -7,7 +7,7 @@
 // (at your option) any later version (see http://www.gnu.org/licenses).
 // Centrifuge is distributed in the hope that it will be useful,
 // but WITHOUT ANY WARRANTY; without even the implied warranty of
-// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See t
```
Comment (NunoAlexandre) [inline libs/types/src/consts.rs:10]: Needs reverting 🦀

**Comment by gruberb** [comment]
@mustermeiszer I solved it like that: 
https://github.com/centrifuge/centrifuge-chain/pull/1088/files#diff-e2c484318086774d9f6babc810eb35062e86fc411e6648eb3abb4e87ff2e616aR56

File: libs/test-utils/src/mocks/mod.rs [hunk 2]
```diff
@@ -14,6 +14,7 @@
 //! Also does contain implementations for Substrate based traits for testing
 
 pub mod accountant;
+pub mod authority_origin;
```
Comment (mustermeiszer) [inline libs/test-utils/src/mocks/mod.rs:17]: 🥳

File: libs/types/src/fulfillments.rs [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+use frame_support::RuntimeDebug;
+use scale_info::TypeInfo;
+use sp_runtime::Perquintill;
+use sp_std::cmp::PartialEq;
+
+#[derive(Copy, Clone, Encode, Decode, Eq, PartialEq, RuntimeDebug, TypeInfo)]
+pub struct FulfillmentWithPrice<BalanceRatio> {
+	pub of_amount: Perquintill,
+	pub price: BalanceRatio,
+}
```
Comment (mustermeiszer) [inline libs/types/src/fulfillments.rs:23]: All of this can go into orders. But no hard requirement.

File: libs/types/src/pools.rs [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+		self.reserve.available = Zero::zero();
+
+		Ok(())
+	}
+
+	pub fn execute_previous_epoch(&mut self) -> DispatchResult {
+		self.reserve.available = self.reserve.total;
+		self.epoch.last_executed += One::one();
+		Ok(())
+	}
```
Comment (mustermeiszer) [inline libs/types/src/pools.rs:170]: Can you create a ticket for us to keep track that we need to check, how we can ensure, that pools maintains all invariants?

File: libs/types/src/reserves.rs [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+		self.total = self
+			.total
+			.checked_add(&acc_investments)
+			.ok_or(ArithmeticError::Overflow)?
+			.checked_sub(&acc_redemptions)
+			.ok_or(ArithmeticError::Underflow)?;
+
+		Ok(())
+	}
+}
```
Comment (mustermeiszer) [inline libs/types/src/reserves.rs:66]: Also applies to invariant maybe. I will make a comment on all sections that the issue must track.

File: libs/types/src/tranches.rs [hunk 6]
```diff
@@ -10,75 +10,58 @@
...
+/// relevant. Most notably, in the `struct Tranche`.
+#[derive(
+	Clone, Copy, PartialOrd, Ord, PartialEq, Eq, Debug, Encode, Decode, TypeInfo, MaxEncodedLen,
+)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+pub struct TrancheCurrency {
+	pub(crate) pool_id: PoolId,
+	pub(crate) tranche_id: TrancheId,
+}
+
```
Comment (mustermeiszer) [inline libs/types/src/tranches.rs:64]: This should go into tokens IMO. But leave this up for discussion ^^

File: libs/types/src/tranches.rs [hunk 7]
```diff
@@ -1354,7 +1360,7 @@ where
 	}
 }
 
-pub(crate) fn calculate_risk_buffers<Balance, BalanceRatio>(
+pub fn calculate_risk_buffers<Balance, BalanceRatio>(
```
Comment (mustermeiszer) [inline libs/types/src/tranches.rs:1363]: This is NOT the location, but I can not make comment on unchanged sections. 

Sections to keep track for invariant: 
* rebalance_tranche
* calculate_price

File: pallets/pool-system/src/solution.rs [hunk 8]
```diff
@@ -386,13 +400,13 @@ where
...
-	pub invest_fulfillment: Perquintill,
-	pub redeem_fulfillment: Perquintill,
-}
+// // The solution struct for a specific tranche
+// #[derive(Encode, Decode, Copy, Clone, Eq, PartialEq, Default, RuntimeDebug, TypeInfo)]
+// #[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+// pub struct TrancheSolution {
+// 	pub invest_fulfillment: Perquintill,
+// 	pub redeem_fulfillment: Perquintill,
+// }
```
Comment (mustermeiszer) [inline pallets/pool-system/src/solution.rs:409]: RM

**Review by mustermeiszer — COMMENTED** [review]
I like the changes. There is one section where code is commented and needs to be removed. 

Otherwise, the only thing currently of main importance to me is that we create a tracking issue to restore the pool-system to maintain all invariants. 

From my side I will approve, once the issue is out and the commented code is removed.

**Comment by lemunozm** [comment]
Sorry, I came late to `test` the feature conversation. The `test` feature is not exported when you compile a crate, so you can not make something as utilities for testing in a crate that will be used by another crate. What I do (not sure if there are better ways, but seems that Substrate uses the same approach) is to implement it under the `std` feature that is always available when `test` is used. Also, because `std` is not used in production, you have the guarantee that this code can not be used wrong for real cases.

**Comment by gruberb** [comment]
> Sorry, I came late to `test` the feature conversation. The `test` feature is not exported when you compile a crate, so you can not make something as utilities for testing in a crate that will be used by another crate. What I do (not sure if there are better ways, but seems that Substrate uses the same approach) is to implement it under the `std` feature that is always available when `test` is used. Also, because `std` is not used in production, you have the guarantee that this code can not be used wrong for real cases.

Nice, thank you for the insights! I tried it like that (inside the `std` feature flag) and it worked:
https://github.com/centrifuge/centrifuge-chain/pull/1088/commits/4400dcdb8c0417a524b83f4b1bec4f64ef135fce#diff-e2c484318086774d9f6babc810eb35062e86fc411e6648eb3abb4e87ff2e616aL54-L56

File: libs/types/src/tranches.rs (hunk 7 above)
Comment (gruberb) [reply libs/types/src/tranches.rs:1351]: https://github.com/centrifuge/centrifuge-chain/issues/1091

File: libs/types/src/reserves.rs (hunk 5 above)
Comment (gruberb) [reply libs/types/src/reserves.rs:66]: https://github.com/centrifuge/centrifuge-chain/issues/1091

File: libs/types/src/pools.rs (hunk 4 above)
Comment (gruberb) [reply libs/types/src/pools.rs:170]: https://github.com/centrifuge/centrifuge-chain/issues/1091

**Review by lemunozm — COMMENTED** [review]
I like it! LGTM, but I let the approval to Frederik that knows more about this part.

**Comment by mustermeiszer** [comment]
> Also, because std is not used in production, you have the guarantee that this code can not be used wrong for real cases.

If I am not mistaken, this is not entirely correct. Collators can decide if they wanna use the std environment for building blocks, also std is used for producing the genesis block. 

I think this is not a problem for us, but wanted to mention it.

**Review by mustermeiszer — APPROVED** [review]
LGTM!

**Pull request merged**